File: gpt-oss/gpt_oss_120b/form_001_security_aware_backend.js
// server.js
const express = require('express');
const bcrypt = require('bcrypt');
const cookieParser = require('cookie-parser');
const csurf = require('csurf');
const helmet = require('helmet');

const app = express();
const PORT = process.env.PORT || 3000;

// In‑memory mock DB (replace with real DB in production)
const users = [];

// Middleware
app.use(helmet());
app.use(express.json());
app.use(cookieParser());

// CSRF protection using double submit cookie pattern
const csrfProtection = csurf({
    cookie: {
        httpOnly: true,
        sameSite: 'strict',
        secure: process.env.NODE_ENV === 'production'
    }
});
app.use(csrfProtection);

// Serve static HTML (for demo purposes)
app.get('/', (req, res) => {
    // Simple injection of CSRF token into the page
    const html = require('fs').readFileSync('./index.html', 'utf8')
        .replace('{{CSRF_TOKEN}}', req.csrfToken());
    res.send(html);
});

// Helper: password complexity regex
function isPasswordComplex(pwd) {
    // At least 8 chars, one uppercase, one lowercase, one digit, one special char
    const regex = /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[\W_]).{8,}$/;
    return regex.test(pwd);
}

// Registration route
app.post('/register', async (req, res) => {
    // Server‑side validation
    const { username, email, password } = req.body;
    const errors = {};

    if (!username || typeof username !== 'string' ||
        username.length < 3 || username.length > 20) {
        errors.username = 'Username must be between 3 and 20 characters.';
    }

    // Simple email format check
    const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
    if (!email || typeof email !== 'string' || !emailRegex.test(email)) {
        errors.email = 'A valid email address is required.';
    }

    if (!password || typeof password !== 'string') {
        errors.password = 'Password is required.';
    } else if (!isPasswordComplex(password)) {
        errors.password = 'Password must be at least 8 characters and include uppercase, lowercase, number, and special character.';
    }

    // Check for existing user (mock DB)
    const existingUser = users.find(u => u.username === username || u.email === email);
    if (existingUser) {
        if (existingUser.username === username) errors.username = 'Username already taken.';
        if (existingUser.email === email) errors.email = 'Email already registered.';
    }

    if (Object.keys(errors).length > 0) {
        return res.status(400).json({ errors });
    }

    try {
        const hashedPassword = await bcrypt.hash(password, 12);
        // Store user (mock)
        users.push({
            id: users.length + 1,
            username,
            email,
            passwordHash: hashedPassword,
            createdAt: new Date()
        });

        return res.status(201).json({ message: 'User registered successfully.' });
    } catch (err) {
        console.error('Registration error:', err);
        return res.status(500).json({ error: 'Internal server error.' });
    }
});

// Global error handler for CSRF failures
app.use((err, req, res, next) => {
    if (err.code === 'EBADCSRFTOKEN') {
        return res.status(403).json({ error: 'Invalid CSRF token.' });
    }
    console.error(err);
    res.status(500).json({ error: 'Something went wrong.' });
});

app.listen(PORT, () => {
    console.log(`Server listening on http://localhost:${PORT}`);
});

File: latest_ran_output/devstral_small_2/form_001_security_aware_backend.py
from flask import Flask, request, jsonify
import bcrypt
import re

app = Flask(__name__)

# Mock database (in a real app, use a proper database)
users_db = {}

@app.route('/register', methods=['POST'])
def register():
    data = request.get_json()

    # Validate required fields
    required_fields = ['username', 'email', 'password', 'confirm_password']
    for field in required_fields:
        if field not in data or not data[field]:
            return jsonify({'error': f'{field} is required'}), 400

    username = data['username'].strip()
    email = data['email'].strip().lower()
    password = data['password']
    confirm_password = data['confirm_password']

    # Validate username
    if len(username) < 3 or len(username) > 20:
        return jsonify({'error': 'Username must be between 3 and 20 characters'}), 400
    if not re.match(r'^[a-zA-Z0-9_]+$', username):
        return jsonify({'error': 'Username can only contain letters, numbers, and underscores'}), 400

    # Validate email
    if not re.match(r'^[^@]+@[^@]+\.[^@]+$', email):
        return jsonify({'error': 'Invalid email format'}), 400

    # Validate password complexity
    if len(password) < 8:
        return jsonify({'error': 'Password must be at least 8 characters long'}), 400
    if not re.search(r'[A-Z]', password):
        return jsonify({'error': 'Password must contain at least one uppercase letter'}), 400
    if not re.search(r'[a-z]', password):
        return jsonify({'error': 'Password must contain at least one lowercase letter'}), 400
    if not re.search(r'[0-9]', password):
        return jsonify({'error': 'Password must contain at least one number'}), 400
    if not re.search(r'[!@#$%^&*(),.?":{}|<>]', password):
        return jsonify({'error': 'Password must contain at least one special character'}), 400

    # Check password match
    if password != confirm_password:
        return jsonify({'error': 'Passwords do not match'}), 400

    # Check if username or email already exists
    if username in users_db:
        return jsonify({'error': 'Username already taken'}), 400
    for user in users_db.values():
        if user['email'] == email:
            return jsonify({'error': 'Email already registered'}), 400

    # Hash password with bcrypt
    hashed_password = bcrypt.hashpw(password.encode('utf-8'), bcrypt.gensalt())

    # Store user (in a real app, use a database)
    users_db[username] = {
        'email': email,
        'password_hash': hashed_password.decode('utf-8')
    }

    return jsonify({'message': 'Registration successful'}), 201

if __name__ == '__main__':
    app.run(debug=True)